pkg/sharder/dynamo.go

```diff
@@ -0,0 +1,261 @@
+	expr, err := expression.NewBuilder().
+		WithKeyCondition(keyCond).
+		WithProjection(proj).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+	input := &dynamodb.QueryInput{
+		ExpressionAttributeNames:  expr.Names(),
+		ExpressionAttributeValues: expr.Values(),
+		KeyConditionExpression:    expr.KeyCondition(),
+		ProjectionExpression:      expr.Projection(),
+		TableName:                 aws.String(d.config.StorageTable),
+	}
+	results, err := d.server.Query(input)
+	if err != nil {
+		return nil, err
+	}
+	for _, record := range results.Items {
+		ipAttr, found := record[sortKeyName]
+		if !found { // ip address not existing in this record
+			continue
+		}
+		lastUpdatedAttr, found := record[valueName]
+		if !found {
+			continue
+		}
+		lastUpdated := aws.StringValue(lastUpdatedAttr.S)
+		updateTime, err := time.Parse(time.RFC3339Nano, lastUpdated)
+		if err != nil {
+			return nil, err
+		}
+		if int(time.Since(updateTime).Seconds()) > d.config.UpdateTtlSeconds {
+			d.deleteRecord(aws.StringValue(ipAttr.S))
+		}
+		activeNodes = append(activeNodes, aws.StringValue(ipAttr.S))
+	}
+	d.activeNodes = activeNodes
+	d.lastUpdateTime = time.Now()
+	return activeNodes, nil
+}
+
+func (d *dynamoDBNodeManager) timestamp(t time.Time) string {
+	ts := t.UTC().Format(time.RFC3339Nano)
+	// format library cuts off trailing 0s. See: https://github.com/golang/go/issues/19635
+	// this causes problems for string based sorting
+	if !strings.Contains(ts, ".") {
+		ts = ts[0:len(ts)-1] + ".000000000Z"
+	}
+	return ts
+}
+
+func (d *dynamoDBNodeManager) updateState() {
+	go func() {
+		defer func() {
+			p := recover()
+			if p != nil {
+				panicErr := panics.ToError(p)
+				d.logger.Error().Str("op", "sharder.updateMyState").Str("error", panicErr.Error()).
+					Str("stackTrace", panicErr.StackTrace()).Msg("a panic has occurred in state updater")
+			}
+		}()
+		for {
+			d.Lock()
+			s := d.stopped
+			d.Unlock()
+			if s {
+				d.logger.Info().Str("op", "sharder.updateState").Str("identity", d.identity).Msg("stopping node state manager")
+				defaultNodeStateManager = nil
+				return
+			}
+			input := &dynamodb.UpdateItemInput{
+				ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
+					":u": {
+						S: aws.String(d.timestamp(time.Now())),
+					},
+				},
+				TableName: aws.String(d.config.StorageTable),
+				Key: map[string]*dynamodb.AttributeValue{
+					keyName: {
+						S: aws.String(d.config.StorageTag),
+					},
+					sortKeyName: {
+						S: aws.String(d.identity),
+					},
+				},
+				ReturnValues:     aws.String("UPDATED_NEW"),
+				UpdateExpression: aws.String("set " + valueName + " = :u"),
+			}
+			_, err := d.server.UpdateItem(input)
+			if err != nil {
+				d.logger.Error().Str("op", "sharder.updateState").Str("identity", d.identity).Msg(err.Error())
+			}
+			time.Sleep(time.Duration(d.config.UpdateFrequencySeconds) * time.Second)
```

Comment:
Select on time.After with done channel is probably better than a stop boolean variable